src/Microsoft.ML.Tokenizers/Model/BPE.cs

```diff
@@ -415,45 +418,59 @@ internal Word MergeWord(string w)
             return word;
         }
 
-        // internal Word.Enumerator WordToTokens(Word word) => word.GetIterator(VocabReverse);
-        internal List<Token> WordToTokens(ref Word word)
+        internal List<Token> WordToTokens(ref Word word) => word.ToTokens(VocabReverse);
+
+        internal List<Token> TokenizeWithCache(string sequence)
         {
-            List<Token> tokens = new(word.SymbolsCount);
+            Word word;
+            if (Cache is not null)
+            {
+                if (Cache.TryGet(sequence, out word))
+                {
+                    return WordToTokens(ref word);
+                }
 
-            foreach (Token token in word.GetIterator(VocabReverse))
+                word = MergeWord(sequence);
+                Cache.Set(sequence, word);
+            }
+            else
             {
-                tokens.Add(token);
+                word = MergeWord(sequence);
             }
 
-            return tokens;
+            return WordToTokens(ref word);
```

Comment:
Can this whole method just be:
```C#
List<Token> result = new();
TokenizeToIdsWithCache(sequence, result);
return result;
```
?

Reply:
No, `TokenizeToIdsWithCache(sequence, result);` fill only the Ids and removing the overhead for filling the whole tokens. Note the difference in the implementation between `TokenizeToIdsWithCache` and `TokenizeWithCache`,. The first is calling `WordToIds` while the second is calling `WordToTokens`